core/transaction_pool/impl/transaction_pool_impl.cpp

```diff
@@ -48,6 +60,53 @@ namespace kagome::transaction_pool {
     metric_ready_txs_->set(0);
   }
 
+  outcome::result<primitives::Transaction>
+  TransactionPoolImpl::constructTransaction(
+      primitives::TransactionSource source,
+      primitives::Extrinsic extrinsic) const {
+    OUTCOME_TRY(res, ttq_->validate_transaction(source, extrinsic));
+
+    return visit_in_place(
+        res,
+        [&](const primitives::TransactionValidityError &e) {
+          return visit_in_place(
+              e,
+              // return either invalid or unknown validity error
+              [](const auto &validity_error)
+                  -> outcome::result<primitives::Transaction> {
+                return validity_error;
+              });
+        },
+        [&](const primitives::ValidTransaction &v)
+            -> outcome::result<primitives::Transaction> {
+          common::Hash256 hash = hasher_->blake2b_256(extrinsic.data);
+          size_t length = extrinsic.data.size();
+
+          return primitives::Transaction{extrinsic,
+                                         length,
+                                         hash,
+                                         v.priority,
+                                         v.longevity,
+                                         v.requires,
+                                         v.provides,
+                                         v.propagate};
+        });
+  }
+
+  outcome::result<Transaction::Hash> TransactionPoolImpl::submitExtrinsic(
+      primitives::TransactionSource source, primitives::Extrinsic extrinsic) {
+    OUTCOME_TRY(tx, constructTransaction(source, extrinsic));
+
+    if (tx.should_propagate) {
+      tx_transmitter_->propagateTransactions(gsl::make_span(std::vector{tx}));
+    }
+    auto hash = tx.hash;
+    // send to pool
+    OUTCOME_TRY(submitOne(std::move(tx)));
+
+    return hash;
+  }
+
   outcome::result<void> TransactionPoolImpl::submitOne(Transaction &&tx) {
     return submitOne(std::make_shared<Transaction>(std::move(tx)));
```

Comment:
Isn't std::move redundant here?

Reply:
It will be copy anyway, because hash in this case is array constructed on stack